### PR TITLE
ScheduledReporter.start() starts with undesired delay

### PR DESCRIPTION
**Pull request by sofax**
`ScheduledReporter.start()` incorrectly uses the interval size for the start delay, too:

```
    public void start(long period, TimeUnit unit) {
        executor.scheduleAtFixedRate(new Runnable() {
            @Override
            public void run() {
                try {
                    report();
                } catch (RuntimeException ex) {
                    LOG.error("RuntimeException thrown from {}#report. Exception was suppressed.", ScheduledReporter.this.getClass().getSimpleName(), ex);
                }
            }
        }, period, period, unit);
    }
```

This is a serious problem when the interval is sufficiently big. I suggest to add an alternative version of `ScheduledReporter.start()` to fix the bug without affecting code that relies on this behavior:

```
    public void start(long period, long initialDelay, TimeUnit unit) {
        executor.scheduleAtFixedRate(new Runnable() {
            @Override
            public void run() {
                try {
                    report();
                } catch (RuntimeException ex) {
                    LOG.error("RuntimeException thrown from {}#report. Exception was suppressed.", ScheduledReporter.this.getClass().getSimpleName(), ex);
                }
            }
        }, initialDelay, period, unit);
    }
```

### CODE REVIEW

File: metrics-core/src/main/java/com/codahale/metrics/ScheduledReporter.java [hunk 1]
```diff
@@ -145,6 +145,26 @@ public void run() {
     }
 
     /**
+     * Starts the reporter polling at the given period.
+     *
+     * @param period       the amount of time between polls
+     * @param initialDelay the time to delay the first execution
+     * @param unit         the unit for {@code period}
+     */
+    public void start(long period, long initialDelay, TimeUnit unit) {
```
Comment (arteam) [inline metrics-core/src/main/java/com/codahale/metrics/ScheduledReporter.java]: I think it would be better to change the order of parameters, so `initialDelay` comes first, to maintain the same order as in the `scheduleAtFixedRate` method.

File: metrics-core/src/main/java/com/codahale/metrics/ScheduledReporter.java [hunk 2]
```diff
@@ -145,6 +145,26 @@ public void run() {
...
+       executor.scheduleAtFixedRate(new Runnable() {
```
Comment (arteam) [inline metrics-core/src/main/java/com/codahale/metrics/ScheduledReporter.java]: I think we could avoid duplication and implement the method as:

```java
synchronized public void start(long initialDelay, long period, TimeUnit unit) {
        if (this.scheduledFuture != null) {
            throw new IllegalArgumentException("Reporter already started");
        }
        this.scheduledFuture = executor.scheduleAtFixedRate(new Runnable() {
           @Override
           public void run() {
               try {
                   report();
               } catch (RuntimeException ex) {
                   LOG.error("RuntimeException thrown from {}#report. Exception was suppressed.", ScheduledReporter.this.getClass().getSimpleName(), ex);
               }
           }
       }, initialDelay, period, unit);
   }
```

as then make the `start(long period, TimeUnit unit)` method to delegate to it as
```java
start(period, period, unit)
```

File: metrics-servlets/src/main/java/com/codahale/metrics/servlets/MetricsServlet.java [hunk 3]
```diff
@@ -64,6 +64,9 @@ protected String getAllowedOrigin() {
 
         /**
          * Returns the name of the parameter used to specify the jsonp callback, if any.
+         *
```
Comment (arteam) [inline metrics-servlets/src/main/java/com/codahale/metrics/servlets/MetricsServlet.java]: It seems to me as an unrelated change. Could you please revert it?

File: metrics-core/src/main/java/com/codahale/metrics/ScheduledReporter.java [hunk 4]
```diff
@@ -145,6 +145,26 @@ public void run() {
     }
 
     /**
+     * Starts the reporter polling at the given period.
```
Comment (arteam) [inline metrics-core/src/main/java/com/codahale/metrics/ScheduledReporter.java]: Could you add a simple unit test to `ScheduledReporterTest`? A simple test which verifies that the scheduler wakes up with a correct initial delay. Very similar to the `pollsPeriodically` test.

File: metrics-servlets/src/main/java/com/codahale/metrics/servlets/MetricsServlet.java (hunk 3 above)
Comment (sofax) [inline metrics-servlets/src/main/java/com/codahale/metrics/servlets/MetricsServlet.java]: Sorry, already working on it - it got pushed too soon.

File: metrics-core/src/main/java/com/codahale/metrics/ScheduledReporter.java (hunk 1 above)
Comment (sofax) [inline metrics-core/src/main/java/com/codahale/metrics/ScheduledReporter.java]: Agreed. :)

File: metrics-core/src/main/java/com/codahale/metrics/ScheduledReporter.java (hunk 2 above)
Comment (sofax) [inline metrics-core/src/main/java/com/codahale/metrics/ScheduledReporter.java]: Again, agreed. 

The lines

```
        if (this.scheduledFuture != null) {
            throw new IllegalArgumentException("Reporter already started");
        }
```

Are new, but they do look reasonable to me.

File: metrics-core/src/main/java/com/codahale/metrics/ScheduledReporter.java (hunk 4 above)
Comment (sofax) [inline metrics-core/src/main/java/com/codahale/metrics/ScheduledReporter.java]: Will do. All those tests are a little bit unreliable though, as the executor services rely on the system time instead of using a dedicated (mockable) clock.

Comment (sofax) [inline metrics-core/src/main/java/com/codahale/metrics/ScheduledReporter.java]: Actually, instead of waiting for the reporter to actually do its work, it would be much better to just verify that the method scheduleAtFixedRate() was called with the correct arguments. There is no need to test the executor itself, as that's a collaborator that has already been tested elsewhere.

Comment (arteam) [inline metrics-core/src/main/java/com/codahale/metrics/ScheduledReporter.java]: > Actually, instead of waiting for the reporter to actually do its work, it would be much better to just verify that the method scheduleAtFixedRate() was called with the correct arguments. There is no need to test the executor itself, as that's a collaborator that has already been tested elsewhere.

Agrred. This would work, too.

Comment (sofax) [inline metrics-core/src/main/java/com/codahale/metrics/ScheduledReporter.java]: Also, the tests are not isolated. The collaborators (mocks) should be initialized per test method, not per class.

Comment (sofax) [inline metrics-core/src/main/java/com/codahale/metrics/ScheduledReporter.java]: Gnnhhh ... my IDE is configured to replace ".*" imports with the explicit ones - but I did NOT want to push those changes, as they were not necessary for the fix.

File: metrics-servlets/src/main/java/com/codahale/metrics/servlets/MetricsServlet.java (hunk 3 above)
Comment (sofax) [inline metrics-servlets/src/main/java/com/codahale/metrics/servlets/MetricsServlet.java]: This change has been reverted.

File: metrics-core/src/main/java/com/codahale/metrics/ScheduledReporter.java (hunk 4 above)
Comment (sofax) [inline metrics-core/src/main/java/com/codahale/metrics/ScheduledReporter.java]: Test cases have been added.

Comment (arteam) [inline metrics-core/src/main/java/com/codahale/metrics/ScheduledReporter.java]: > Gnnhhh ... my IDE is configured to replace ".*" imports with the explicit ones - but I did NOT want to push those changes, as they were not necessary for the fix.

I think expanding star imports is a good change, we can leave it.

**Comment by arteam** [comment]
Thank for your contribution!
